src/serialization.rs

```diff
@@ -0,0 +1,270 @@
+                    dead_letter: DeadLetter::from_failed_deserialization(payload, e.to_string()),
+                });
+            }
+        };
+
+        Ok(value)
+    }
+}
+
+struct AvroDeserializer {
+    decoder: EasyAvroDecoder,
+}
+
+#[derive(Default)]
+struct AvroSchemaDeserializer {
+    schema: Option<apache_avro::Schema>,
+}
+
+struct JsonDeserializer {
+    decoder: EasyJsonDecoder,
+}
+
+#[async_trait]
+impl MessageDeserializer for AvroDeserializer {
+    async fn deserialize(
+        &mut self,
+        message_bytes: &[u8],
+    ) -> Result<Value, MessageDeserializationError> {
+        match self.decoder.decode_with_schema(Some(message_bytes)).await {
+            Ok(drs) => match drs {
+                Some(v) => match Value::try_from(v.value) {
+                    Ok(v) => Ok(v),
+                    Err(e) => Err(MessageDeserializationError::AvroDeserialization {
+                        dead_letter: DeadLetter::from_failed_deserialization(
+                            message_bytes,
+                            e.to_string(),
+                        ),
+                    }),
+                },
+                None => return Err(MessageDeserializationError::EmptyPayload),
+            },
+            Err(e) => {
+                return Err(MessageDeserializationError::AvroDeserialization {
+                    dead_letter: DeadLetter::from_failed_deserialization(
+                        message_bytes,
+                        e.to_string(),
+                    ),
+                });
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl MessageDeserializer for AvroSchemaDeserializer {
+    async fn deserialize(
+        &mut self,
+        message_bytes: &[u8],
+    ) -> Result<Value, MessageDeserializationError> {
+        let reader_result = match &self.schema {
+            None => apache_avro::Reader::new(Cursor::new(message_bytes)),
+            Some(schema) => apache_avro::Reader::with_schema(schema, Cursor::new(message_bytes)),
+        };
+
+        match reader_result {
+            Ok(mut reader) => {
+                if let Some(r) = reader.next() {
+                    let v = match r {
+                        Err(_) => return Err(MessageDeserializationError::EmptyPayload),
+                        Ok(v) => Value::try_from(v),
+                    };
+
+                    return match v {
+                        Ok(value) => Ok(value),
+                        Err(e) => Err(MessageDeserializationError::AvroDeserialization {
+                            dead_letter: DeadLetter::from_failed_deserialization(
+                                message_bytes,
+                                e.to_string(),
+                            ),
+                        }),
+                    };
+                }
+
+                return Err(MessageDeserializationError::EmptyPayload);
+                // TODO: Code to return multiple values from avro message
```

Comment:
Should this TODO block the merge of this code?

Reply:
That todo is there to prompt discussion. Right now KDI is designed to process one message at a time. Avro allows multiple records within the message. It's a pretty large shift in the way we do things. My original `MessageDeserializer` returned `(Vec<Value>, u64)` to account for that. However, I ultimately left it as a todo so that we could give it some thought without this PR radically changing the way the process works.